hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java

```diff
@@ -311,21 +321,27 @@ long newExecutionId() {
      * @throws IllegalStateException if the JobResult is already present
      */
     void completeJob(
-            long jobId,
+            @Nonnull MasterContext masterContext,
             @Nullable List<RawJobMetrics> terminalMetrics,
-            @Nonnull String coordinator,
             long completionTime,
             @Nullable Throwable error
     ) {
+        long jobId = masterContext.jobId();
+
         JobRecord jobRecord = getJobRecord(jobId);
         if (jobRecord == null) {
             throw new JobNotFoundException(jobId);
         }
 
+        DAG dag = deserializeDag(masterContext, jobRecord);
```

Comment:
deserializing the DAG is expensive operation. We did a lot of optimization so that there is as little overhead as possible during Job Completion and this undoes it. We can know which observables are known in MasterContext during job initialization without having to deserialize it again.